src/payload_storage.rs

```diff
@@ -181,24 +183,144 @@ impl PayloadStorage {
-            let slot_id = page.write().insert_value(&comp_payload).unwrap();
+
+            let slot_id = page
+                .write()
+                .insert_value(point_offset, &comp_payload)
+                .unwrap();
+
             // update page_tracker
             self.page_tracker
+                .write()
                 .set(point_offset, PagePointer::new(page_id, slot_id));
         }
     }
 
     /// Delete a payload from the storage
     /// Returns None if the point_offset, page, or payload was not found
     pub fn delete_payload(&mut self, point_offset: PointOffset) -> Option<()> {
-        let PagePointer { page_id, slot_id } = self.get_pointer(point_offset).copied()?;
+        let PagePointer { page_id, slot_id } = self.get_pointer(point_offset)?;
         let page = self.pages.get_mut(&page_id)?;
         // delete value from page
         page.write().delete_value(slot_id);
         // delete mapping
-        self.page_tracker.unset(point_offset);
+        self.page_tracker.write().unset(point_offset);
         Some(())
     }
+
+    /// Page ids with amount of fragmentation, ordered by most to least fragmentation
+    fn pages_to_defrag(&self) -> Vec<(u32, usize)> {
+        let mut fragmentation = self
+            .pages
+            .iter()
+            .filter_map(|(page_id, page)| {
+                let page = page.read();
+                let frag_space = page.fragmented_space();
+
+                // check if we should defrag this page
+                let frag_threshold =
+                    SlottedPageMmap::FRAGMENTATION_THRESHOLD_RATIO * page.page_size() as f32;
+                if frag_space < frag_threshold.ceil() as usize {
+                    // page is not fragmented enough, skip
+                    return None;
+                }
+                Some((*page_id, frag_space))
+            })
+            .collect::<Vec<_>>();
+
+        // sort by most to least fragmented
+        fragmentation.sort_unstable_by_key(|(_, fragmented_space)| Reverse(*fragmented_space));
+
+        fragmentation
+    }
+
+    pub fn compact(&mut self) {
+        // find out which pages should be compacted
+        let pages_to_defrag = self.pages_to_defrag();
+
+        if pages_to_defrag.is_empty() {
+            return;
+        }
+
+        let mut pages_to_defrag = pages_to_defrag.into_iter();
+        let mut old_page_id = pages_to_defrag.next().unwrap().0;
+        let mut last_slot_id = 0;
+
+        // TODO: account for the fact that the first value could be larger than 32MB and the newly created page will
+        // immediately not be used? we don't want to create empty pages. But it is a quite rare case, so maybe we can just ignore it
+        let mut size_hint = None;
+
+        // This is a loop because we might need to create more pages if the current new page is full
+        'new_page: loop {
+            // create new page
+            let new_page_id = self.create_new_page(size_hint);
```

Comment:
I wonder if we should create this page as a `tempfile` and then move it at the end of compaction.
This way there is no left over if compaction is interrupted.